tokio/src/fs/file.rs

```diff
@@ -538,6 +538,172 @@ impl File {
     pub fn set_max_buf_size(&mut self, max_buf_size: usize) {
         self.max_buf_size = max_buf_size;
     }
+
+    /// Reads a number of bytes starting from a given offset.
+    ///
+    /// Returns the number of bytes read.
+    ///
+    /// The offset is relative to the start of the file and thus independent
+    /// from the current cursor.
+    ///
+    /// The current file cursor is not affected by this function.
+    ///
+    /// It is not an error to return with a short read.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// use tokio::fs::File;
+    /// use tokio::io::AsyncSeekExt;
+    ///
+    /// # async fn dox() -> std::io::Result<()> {
+    /// let mut buf = vec![0_u8; 10];
+    /// let mut file = File::open("foo.txt").await?;
+    /// file.read_at(&mut buf, 5).await?;
+    ///
+    /// assert_eq!(file.stream_position().await?, 0);
+    /// # Ok(())
+    /// # }
+    /// ```
+    #[cfg(unix)]
+    #[cfg_attr(docsrs, doc(cfg(unix)))]
+    pub async fn read_at(&self, buf: &mut [u8], offset: u64) -> io::Result<usize> {
+        fn _read_at(std: &StdFile, n: usize, offset: u64) -> io::Result<Vec<u8>> {
+            use std::os::unix::fs::FileExt;
+            let mut buf: Vec<u8> = vec![0; n];
+            let n_read = std.read_at(&mut buf, offset)?;
+            buf.truncate(n_read);
+
+            Ok(buf)
+        }
+
+        let std = self.std.clone();
+        let n = buf.len();
+        let bytes_read = asyncify(move || _read_at(&std, n, offset)).await?;
+        let len = bytes_read.len();
+        buf[..len].copy_from_slice(&bytes_read);
```

Comment:
The Tokio file already has a bunch of logic for keeping track of and reusing a buffer, but none of these functions use that logic. Is there a reason you're doing it this way?

Reply:
> The Tokio file already has a bunch of logic for keeping track of and reusing a buffer,

Well, I am not aware of this, do you mean the `bytes` crate? If not, would you like to show me some examples?

Reply:
No, I mean the stuff related to these types:

https://github.com/tokio-rs/tokio/blob/c98e22fc5a275e083c3079424bfbf964fba0f7a9/tokio/src/fs/file.rs#L96-L118

Reply:
Thanks for showing me this!

I just took another look at other methods implemented with `spawn_blocking()`, seems that I should add:

```rs
        let mut inner = self.inner.lock().await;
        inner.complete_inflight().await;
```

before involving the actual logic, right?

Reply:
One could probably argue either way about whether it's even desireable. One advantage of `read_at` is that you can have several calls in parallel, but if we use the state logic, then they will run one-after-the-other.

I'm not sure what the best answer here is.

Reply:
> One advantage of read_at is that you can have several calls in parallel, but if we use the state logic, then they will run one-after-the-other.

Thanks for pointing it out! [The capability of enabling concurrent access is indeed the reason why `pread/pwrite` are added](https://stackoverflow.com/a/5268578/14092446), so I slightly tend to use separate buffers

Reply:
> One could probably argue either way about whether it's even desireable. One advantage of `read_at` is that you can have several calls in parallel, but if we use the state logic, then they will run one-after-the-other.
> 
> I'm not sure what the best answer here is.

I have a tentative plan, can we use `RwLock` instead of `Mutex` to achieve sharing among multiple readers and allowing them run in parallel?

https://github.com/tokio-rs/tokio/blob/c98e22fc5a275e083c3079424bfbf964fba0f7a9/tokio/src/fs/file.rs#L92

Reply:
@Chasing1020 A read/write lock doesn't help. Fundamentally, the shared buffer can only hold one piece of data at the time.